master/internal/rm/kubernetesrm/request_queue.go

```diff
@@ -231,15 +229,15 @@ func (r *requestQueue) receiveWorkerAvailable(ctx *actor.Context, msg workerAvai
 		// If both creation and deletion are nil it means that the creation
 		// request was canceled.
 		if nextRequest.createResources != nil {
-			delete(r.pendingResourceCreations, nextRequest.createResources.handler)
-			r.creationInProgress.Insert(nextRequest.createResources.handler)
-			ctx.Tell(ctx.Sender(), *nextRequest.createResources)
+			next := getCallbackPtr(nextRequest.createResources.errorHandler)
+			delete(r.pendingResourceCreations, next)
+			r.creationInProgress.Insert(next)
+			r.workerChan <- *nextRequest.createResources
 			return
 		} else if nextRequest.deleteResources != nil {
-			ctx.Tell(ctx.Sender(), *nextRequest.deleteResources)
+			r.workerChan <- *nextRequest.deleteResources
 			return
 		}
 	}
-
-	r.availableWorkers = append(r.availableWorkers, ctx.Sender())
+	r.availableWorkers++
```

Comment:
i think the existing code has this strange behavior around tracking available workers and distributing work only when it is available because it is impossible to pluck messages out of an actor inbox if, e.g., a delete comes in with a create in the queue.

i think we should aim to remove `workerAvailableCallback`/`availableWorkers` completely and use a map+slice (or the ordered map construct) to track and abort in flight work. the number of callbacks and how they are being passed around, callbacks of callbacks, and strange types (`getCallbackPtr`) make this harder to follow than the actor code in my opinion.

Reply:
Sure thing, I am looking into making it less complicated by using interfaces instead of function callbacks from our conversation this morning. The availableWorkers variable is pretty much just used for testing at this point and I can take a look at removing it. Was hoping we could resolve the ordered map thing earlier but perhaps we can talk again about a solution.

Reply:
let's just chat again if you want, now that i got a chance to digest this